traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/dns/ZoneManager.java

```diff
@@ -460,18 +499,37 @@ private static void generateZones(final TrafficRouter tr, final LoadingCache<Zon
 			LOGGER.fatal("Unable to create zone: " + ex.getMessage(), ex);
 		}
 
-		primeZoneCache(domain, name, list, tr, zc, dzc, generationTasks, primingTasks, ds);
+		primeZoneCache(domain, name, list, tr, zc, dzc, generationTasks, primingTasks, ds, newDomainsToZoneKeys);
 
 		return records;
 	}
 
-	@SuppressWarnings("PMD.CyclomaticComplexity")
+	@SuppressWarnings({"PMD.CyclomaticComplexity", "PMD.ExcessiveParameterList"})
 	private static void primeZoneCache(final String domain, final Name name, final List<Record> list, final TrafficRouter tr,
 			final LoadingCache<ZoneKey, Zone> zc, final LoadingCache<ZoneKey, Zone> dzc, final List<Runnable> generationTasks,
-			final BlockingQueue<Runnable> primingTasks, final DeliveryService ds) {
+			final BlockingQueue<Runnable> primingTasks, final DeliveryService ds, final ConcurrentMap<String, ZoneKey> newDomainsToZoneKeys) {
 		generationTasks.add(() -> {
 			try {
-				final Zone zone = zc.get(signatureManager.generateZoneKey(name, list)); // cause the zone to be loaded into the new cache
+				final ZoneKey newZoneKey = signatureManager.generateZoneKey(name, list);
+				if (tr.isDnssecZoneDiffingEnabled() && domainsToZoneKeys.containsKey(domain)) {
```

Comment:
Sure, I can try that out.

Reply:
So, I tried it to split this out, but I wanted to include all of L513-532 since I wanted all of the modification of `newDomainsToZoneKeys` to be done in the same place. Then I found that the early return on L522 makes splitting this out trickier because the new method would have to have a `Zone` passed in to be set as one method result but also return a `boolean` for whether or not an existing zone was matched and reused.

That said, I think the way the code is currently would be easier to comprehend than what I just described. What do you think?

Reply:
The new method would not need a `Zone` passed in if you are okay also calling declaring setting `Zone zone = zc.get(newZoneKey)` outside that method [for line 541](https://github.com/apache/trafficcontrol/blob/afc6baab28/traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/dns/ZoneManager.java#L541).

The early return on L522 doesn't need to make splitting this out trickier if you are okay having multiple returns. Something like:

```java
private static boolean diffZones(...) throws ExecutionException {
    ...
    if (!tr.isDnssecZoneDiffingEnabled() || !domainsToZoneKeys.containsKey(domain)) {
        return false;
    }
    ...
    if (!zonesAreEqual(newZoneKey.getRecords(), oldZoneKey.getRecords())) {
        ...
        return false;
    }
    ...
    if (oldZone != null) {
        ...
    }
    ...
    return true;
}
```

Reply:
That approach would also involve passing in `newZoneKey`.

Reply:
But yeah, if that complicates it, I agree it's better to leave as-is.

Reply:
Yeah, the thing is `zc.get(newZoneKey)` is what actually creates and signs the zone if the value isn't there already. So I'm pretty sure I have to `newDomainsToZoneKeys.put(domain, newZoneKey);` _after_ doing `zc.get(...)` so that it's definitely already signed if I want to reuse it. I might be wrong about that, since the reference shouldn't change after adding it to `newDomainsToZoneKeys`, but I feel better about not saving it into the map until it's signed for sure.

Reply:
I agree about calling `zc.get(...)` before `newDomainsToZoneKeys.put(domain, newZoneKey)`. As I understand it, the reason you could additionally declare `Zone zone = zc.get(newZoneKey)` after `newDomainsToZoneKeys.put(...)` is that LoadingCache only creates signs the zone if the ZoneKey isn't already there. If it is is, you get the existing zone for free.

Calling `zc.get(newZoneKey)` again doesn't refresh the ZoneKey.

Reply:
Yeah, I think I will just keep the block as-is because I'd rather not add extra `zc.gets` and multi-returns just to remove a level of indentation. In this case I think one extra level of indentation is fine, especially since it's a relatively small block. Once the zone-diffing feature becomes vetted I would like to come back here and remove the feature flag for it, and just always do zone-diffing (with or without dnssec enabled). That will remove a level of indentation here in the future.

Reply:
Fair enough, sounds good to me.